code/iaas/logic/src/main/java/io/cattle/platform/process/externalevent/ExternalServiceEventManagerImpl.java

```diff
@@ -0,0 +1,199 @@
+
+    @Inject
+    LockManager lockManager;
+
+    @Inject
+    DynamicSchemaDao schemaDao;
+
+    @Inject
+    ObjectProcessManager objectProcessManager;
+
+    @Inject
+    ObjectManager objectManager;
+
+    public void handleServiceEvent(final ExternalEvent event, final ProcessState state, final ProcessInstance process) {
+        lockManager.lock(new ExternalEventLock(SERVICE_LOCK_NAME, event.getAccountId(), event.getExternalId()), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                Map<String, Object> serviceData = CollectionUtils.toMap(DataUtils.getFields(event).get(FIELD_SERVICE));
+                if (serviceData.isEmpty()) {
+                    log.warn("Empty service for externalServiceEvent: {}.", event);
+                    return;
+                }
+
+                String kind = serviceData.get(ObjectMetaDataManager.KIND_FIELD) != null ? serviceData.get(ObjectMetaDataManager.KIND_FIELD).toString() : null;
+                if (StringUtils.isEmpty(kind) || schemaDao.getSchema(kind, event.getAccountId()) == null) {
+                    log.warn("Couldn't find schema for service type [{}]. Returning.", kind);
+                    return;
+                }
+
+                if (StringUtils.equals(event.getEventType(), TYPE_SERVICE_CREATE)) {
+                    createService(event, serviceData);
+                } else if (StringUtils.equals(event.getEventType(), TYPE_SERVICE_UPDATE)) {
+                    updateService(event, serviceData);
+                } else if (StringUtils.equals(event.getEventType(), TYPE_SERVICE_DELETE)) {
+                    deleteService(event, serviceData);
+                }
+            }
+        });
+    }
+
+    void createService(ExternalEvent event, Map<String, Object> serviceData) {
+        Service svc = serviceDao.getServiceByExternalId(event.getAccountId(), event.getExternalId());
+        if (svc != null) {
+            return;
+        }
+
+        Environment environment = getEnvironment(event);
+        if (environment == null) {
+            log.info("Can't process service event. Could not get or create environment. Event: [{}]", event);
+            return;
+        }
+
+        Map<String, Object> service = new HashMap<String, Object>();
+        service.put(ObjectMetaDataManager.ACCOUNT_FIELD, event.getAccountId());
+        service.put(FIELD_ENVIRIONMENT_ID, environment.getId());
+        if (serviceData != null) {
+            service.putAll(serviceData);
+        }
+
+        try {
+            String create = objectProcessManager.getStandardProcessName(StandardProcess.CREATE, Service.class);
+            String activate = objectProcessManager.getStandardProcessName(StandardProcess.ACTIVATE, Service.class);
+            ProcessUtils.chainInData(service, create, activate);
+            resourceDao.createAndSchedule(Service.class, service);
+        } catch (ProcessCancelException e) {
```

Comment:
What is the impact of processCancelException (event not being processed)? Do you think we should log it?